src/track/beats.h

```diff
@@ -1,177 +1,239 @@
 };
 
-// Beats is a pure abstract base class for BPM and beat management classes. It
-// provides a specification of all methods a beat-manager class must provide, as
-// well as a capability model for representing optional features.
-class Beats : public QObject {
+/// Beats is a class for BPM and beat management classes.
+/// It stores beats information including beats position, down beats position,
+/// phrase beat position and changes in tempo.
+class Beats final : public QObject {
     Q_OBJECT
+  private:
+    explicit Beats(const audio::StreamInfo& streamInfo = audio::StreamInfo(),
+            const BeatsInternal& internal = BeatsInternal());
+
   public:
-    Beats() { }
     ~Beats() override = default;
+    Beats(const Beats& other) = delete;
+
+    /// The source of this byte array is the serialized representation of beats
+    /// generated by the protocol buffer and stored in the database.
+    void initWithProtobuf(const QByteArray& byteArray);
+    /// A list of beat locations in audio frames may be provided.
+    /// The source of this data is the analyzer.
+    void initWithAnalyzer(const QVector<FramePos>& beats,
+            const QVector<track::io::TimeSignatureMarker>&
+                    timeSignatureMarkers =
+                            QVector<track::io::TimeSignatureMarker>());
+
+    /// Serializes into a protobuf.
+    QByteArray toProtobuf() const;
+
+    /// Returns a string representing the version of the beat-processing code that
+    /// produced this Beats instance. Used by BeatsFactory for associating a
+    /// given serialization with the version that produced it.
+    QString getVersion() const;
+
+    /// Return a string that represent the preferences used to generate
+    /// the beats object.
+    QString getSubVersion() const;
+
+    void setSubVersion(const QString& subVersion);
+
+    /// Initializes the BeatGrid to have a BPM of dBpm and the first beat offset
+    /// of firstBeatFrame. Does not generate an updated() signal, since it is
+    /// meant for initialization.
+    void setGrid(Bpm dBpm, FramePos firstBeatFrame = kStartFramePos);
+
+    /// Starting from frame, return the next beat
+    /// in the track, or invalid beat if none exists. If frame refers to the location
+    /// of a beat, the same beat is returned.
+    Beat findNextBeat(FramePos frame) const;
+
+    /// Starting from frame frame, return the previous
+    /// beat in the track, or invalid beat if none exists. If frame refers to the
+    /// location of beat, the same beat is returned.
+    Beat findPrevBeat(FramePos frame) const;
+
+    /// Starting from frame, fill the frame numbers of the previous beat
+    /// and next beat.  Either can be -1 if none exists.  If frame refers
+    /// to the location of the beat, the first value is frame, and the second
+    /// value is the next beat position.  Non- -1 values are guaranteed to be
+    /// even.  Returns false if *at least one* frame is -1.  (Can return false
```

Comment:
```suggestion
    /// value is the next beat position. Returns false if *at least one* frame is -1.  (Can return false
```
"Non- -1 values a guaranteed to be even" is legacy from when this returned stereo samples and is no longer true.